casper/src/main/scala/coop/rchain/casper/SafetyOracle.scala

```diff
@@ -48,120 +49,172 @@ object SafetyOracle extends SafetyOracleInstances {
+                                          candidates: Map[Validator, Int]): F[Int] = {
+        // TODO: Replace with findM when it gets merged into cats
+        def filterAgreeingJustificationHashes(justificationHashes: List[BlockHash],
+                                              validator: Validator): F[List[BlockHash]] =
+          justificationHashes.filterA(justificationHash =>
             for {
-              firstLatestHash <- blockDag.latestMessages.get(first).toList
-              firstLatest     = internalMap(firstLatestHash)
-              justification <- firstLatest.justifications.map {
-                                case Justification(_, latestBlock: BlockHash) => latestBlock
-                              }
-              justificationBlock <- internalMap.get(justification).toList
-              child              <- filterChildren(justificationBlock, internalMap)
-              if child.sender == second
-            } yield child
-          potentialDisagreements.forall { potentialDisagreement =>
-            compatible(internalMap, estimate, potentialDisagreement)
+              justificationBlock <- unsafeGetBlock[F](justificationHash)
+              isSenderSecond     = justificationBlock.sender == validator
+              compatible         <- computeCompatibility(estimate, justificationBlock)
+            } yield isSenderSecond && compatible)
+
+        def seesAgreement(first: Validator, second: Validator): F[Boolean] = {
+          val maybeFirstLatestHash = blockDag.latestMessages.get(first)
+          maybeFirstLatestHash match {
+            case Some(firstLatestHash) =>
+              for {
+                firstLatestBlock    <- unsafeGetBlock[F](firstLatestHash)
+                justificationHashes = firstLatestBlock.justifications.map(_.latestBlockHash)
+                agreeingJustificationHashes <- filterAgreeingJustificationHashes(
+                                                justificationHashes.toList,
+                                                second)
+              } yield agreeingJustificationHashes.nonEmpty
+            case None => false.pure[F]
           }
         }
 
-        val edges = (for {
-          x <- candidates.keys
-          y <- candidates.keys
-          if x.toString > y.toString // TODO: Order ByteString
-        } yield (x, y)) filter {
-          case (first: Validator, second: Validator) =>
-            seesAgreement(first, second) && seesAgreement(second, first) &&
-              neverEventuallySeeDisagreement(first, second) && neverEventuallySeeDisagreement(
-              second,
-              first)
+        // TODO: Potentially replace with isInBlockDAG
+        // TODO: Replace with isInJustificationChain
+        def filterChildren(candidate: BlockMessage, validator: Validator): F[List[BlockMessage]] =
+          for {
+            internalMap <- BlockStore[F].asMap()
+            children <- internalMap.values.toList.filterA { potentialChild =>
```

Comment:
Rather than looking through literally every block to find the ones created by this validator after the candidate, how about you start with that validator's latest block and use `bfTraverseF` with the justifications filtered by that validator. This would give you a stream of blocks created by that validator in reverse order, so you just stop the stream when you reach `candidate`.

Reply:
Done